Lib/test/pickletester.py

```diff
@@ -2994,6 +2995,59 @@ def __reduce__(self):
 class BBB(object):
     pass
 
+
+class AbstractHookTests(unittest.TestCase):
+    def test_pickler_hook(self):
+        # test the CPickler ability to register custom, user-defined reduction
+        # callbacks to pickle functions and classes.
+
+        def custom_reduction_callback(pickler, obj):
+            obj_name = obj.__name__
+
+            if obj_name == 'f':
+                # asking the pickler to save f as 5
+                return int, (5, )
+
+            if obj_name == 'MyClass':
+                return str, ('NewClass', )
+
+            elif obj_name == 'log':
+                # letting the pickler falling back to buitlin save_global to
+                # pickle functions named 'log' by attribute.
+                return NotImplementedError
+
+            elif obj_name == 'g':
+                # in this case, the callback returns an invalid result (not a
+                # 2-5 tuple), the pickler should raise a proper error.
+                return False
+            return NotImplementedError
+
```

Comment:
Maybe add another case where `custom_reduction_callback` raises an exception that is not `NotImplementedError` (for instance a `ValueError` instance) and check that it is raised back to the call of `p.dump`.